File: backend/src/routes/prediction.js
const express = require('express');
const router  = express.Router();
const axios   = require('axios');
const { verifyToken } = require('../middleware/authMiddleware');
const { computeRiskScore } = require('../utils/riskScore');

// POST /api/predict — predict without saving
router.post('/predict', verifyToken, async (req, res) => {
  try {
    const { heart_rate, temperature, oxygen } = req.body;
    if (!heart_rate || !temperature || !oxygen) {
      return res.status(400).json({ error: 'heart_rate, temperature, oxygen are required.' });
    }
    const result = await runPrediction(req.body);
    res.json(result);
  } catch (error) {
    res.status(500).json({ error: 'Prediction failed', message: error.message });
  }
});

// POST /api/explain — SHAP feature contributions for a prediction
router.post('/explain', verifyToken, async (req, res) => {
  try {
    const { heart_rate, temperature, oxygen } = req.body;
    if (!heart_rate || !temperature || !oxygen) {
      return res.status(400).json({ error: 'heart_rate, temperature, oxygen are required.' });
    }
    const result = await runExplanation(req.body);
    res.json(result);
  } catch (error) {
    res.status(500).json({ error: 'Explanation failed', message: error.message });
  }
});

// Tries the Python SHAP service; falls back to a rule-based contribution estimate.
async function runExplanation(data) {
  if (process.env.ML_SERVICE_URL) {
    try {
      const response = await axios.post(
        `${process.env.ML_SERVICE_URL}/explain`, data, { timeout: 8000 }
      );
      return response.data;
    } catch (mlError) {
      console.warn('SHAP service unavailable, using rule-based explanation:', mlError.message);
    }
  }
  return ruleBasedExplanation(data);
}

// Degraded explanation when the SHAP service is offline: derive per-vital
// contributions from the same risk-score breakdown the app already uses, so
// the UI always shows *why* a risk level was reached.
function ruleBasedExplanation(data) {
  const hr   = parseFloat(data.heart_rate);
  const temp = parseFloat(data.temperature);   // °F (frontend sends °C for ML; rule-based expects °F)
  const spo2 = parseFloat(data.oxygen);
  const fall = data.fall_detected === true || data.fall_detected === 'true' || data.fall_detected === 1;

  // The risk util expects °F. The /explain frontend payload sends °C for the ML
  // model, so convert back to °F here if the value looks like Celsius (< 60).
  const tempF = temp < 60 ? (temp * 9 / 5) + 32 : temp;

  const risk = computeRiskScore({ heart_rate: hr, temperature: tempF, oxygen: spo2, fall_detected: fall });
  const b = risk.breakdown;

  // Positive shap = increases risk. Healthy vitals (0 points) get a small
  // negative value so they read as "decreases risk".
  const mk = (feature, value, points) => ({
    feature,
    value: Math.round(value * 100) / 100,
    shap: points > 0 ? Math.round((points / 100) * 1000) / 1000 : -0.02,
    direction: points > 0 ? 'increases' : 'decreases',
  });

  const contributions = [
    mk('Heart Rate',        hr,    b.heart_rate),
    mk('Oxygen Saturation', spo2,  b.oxygen),
    mk('Body Temperature',  tempF, b.temperature),
  ];
  if (fall) contributions.push({ feature: 'Fall Detected', value: 1, shap: Math.round((b.fall / 100) * 1000) / 1000, direction: 'increases' });
  contributions.sort((a, c) => Math.abs(c.shap) - Math.abs(a.shap));

  const prediction = risk.score >= 67 ? 'Emergency' : risk.score >= 34 ? 'Warning' : 'Normal';

  return {
    prediction,
    raw_label: risk.score >= 34 ? 'High Risk' : 'Low Risk',
    probability_high_risk: Math.min(0.99, risk.score / 100),
    base_value: 0,
    model_used: 'Rule-based (SHAP service offline)',
    contributions,
  };
}

// ── Core prediction: tries ML service, falls back to rule-based ───────────────
async function runPrediction(data) {
  if (process.env.ML_SERVICE_URL) {
    try {
      const response = await axios.post(
        `${process.env.ML_SERVICE_URL}/predict`, data, { timeout: 5000 }
      );
      return response.data;
    } catch (mlError) {
      console.warn('ML service unavailable, using rule-based fallback:', mlError.message);
    }
  }
  return ruleBased(data);
}

// ── Rule-based fallback (matches your model's logic) ─────────────────────────
function ruleBased({ heart_rate, temperature, oxygen, fall_detected,
                     systolic_bp = 120, diastolic_bp = 80 }) {
  const hr   = parseFloat(heart_rate);
  const temp = parseFloat(temperature);
  const spo2 = parseFloat(oxygen);
  const sbp  = parseFloat(systolic_bp);
  let prediction = 'Normal';
  let reasons = [];

  if ((hr > 120 && spo2 < 92) || fall_detected || (temp > 103.5) || sbp > 180) {
    prediction = 'Emergency';
    if (hr > 120 && spo2 < 92) reasons.push('Critical: high heart rate + low oxygen');
    if (fall_detected)          reasons.push('Fall detected');
    if (temp > 103.5)           reasons.push('Dangerously high temperature');
    if (sbp > 180)              reasons.push('Hypertensive crisis');
  } else if (
    (temp > 100.4 && hr > 100) || hr > 110 || hr < 50 ||
    spo2 < 95 || temp > 102 || sbp > 140
  ) {
    prediction = 'Warning';
    if (temp > 100.4) reasons.push('Fever detected');
    if (hr > 110)     reasons.push('High heart rate (tachycardia)');
    if (hr < 50)      reasons.push('Low heart rate (bradycardia)');
    if (spo2 < 95)    reasons.push('Low oxygen saturation');
    if (sbp > 140)    reasons.push('Elevated blood pressure');
  } else {
    reasons.push('All vitals within normal range');
  }

  const recommendations = {
    Normal: 'Continue regular monitoring. Stay hydrated — drink 8 glasses of water daily. Walk at least 20 minutes today. Aim for 7–8 hours of sleep tonight.',
    Warning: 'Rest and avoid strenuous activity. Re-measure vitals in 30 minutes. Contact your doctor if symptoms worsen. ⚠️ Informational only — please consult a healthcare professional.',
    Emergency: '🚨 Seek immediate medical attention. Call emergency services now. Do not stay alone. Do not drive yourself. ⚠️ Informational only — call emergency services immediately.'
  };

  return {
    prediction,
    raw_label    : prediction === 'Normal' ? 'Low Risk' : 'High Risk',
    recommendation: recommendations[prediction],
    reasons,
    model_used   : 'Rule-Based Fallback',
    probability_high_risk: prediction === 'Normal' ? 0.1 : prediction === 'Warning' ? 0.6 : 0.9,
  };
}

module.exports = router;
module.exports.runPrediction = runPrediction;

File: ml-model/src/prediction.py
"""
ML Prediction Microservice — uses YOUR trained RF + XGBoost models
Run: python prediction.py
Listens on: http://localhost:5001
Requirements: pip install flask scikit-learn xgboost numpy joblib pandas
"""
from flask import Flask, request, jsonify
import numpy as np
import os, sys

app = Flask(__name__)

# ── Load models via your RiskPredictor class ──────────────────────────────────
MODEL_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, MODEL_DIR)

try:
    from predict import RiskPredictor
    predictor = RiskPredictor(base_dir=MODEL_DIR)
    print("✅ Models loaded via RiskPredictor")
except Exception as e:
    print(f"⚠️  Could not load models: {e}")
    predictor = None

# ── Recommendation map (3-level from API input) ───────────────────────────────
RECOMMENDATIONS = {
    "Low Risk": (
        "Continue regular monitoring. "
        "Stay hydrated — drink 8 glasses of water daily. "
        "Walk at least 20 minutes today. "
        "Aim for 7–8 hours of sleep tonight."
    ),
    "High Risk": (
        "⚠️ Your vitals indicate a High Risk condition. "
        "Rest immediately and avoid any strenuous activity. "
        "Re-measure your vitals within 30 minutes. "
        "Contact your doctor or seek medical attention if symptoms persist or worsen. "
        "This is informational only — please consult a healthcare professional."
    )
}

def derive_features(data: dict) -> dict:
    """Auto-derive BMI, MAP, Pulse Pressure, HRV from raw vitals."""
    weight  = float(data.get("weight_kg", 70))
    height  = float(data.get("height_m") or data.get("height_cm", 170) / 100)
    sbp     = float(data.get("systolic_bp", 120))
    dbp     = float(data.get("diastolic_bp", 80))
    hr      = float(data.get("heart_rate", 75))

    bmi             = weight / (height ** 2)
    pulse_pressure  = sbp - dbp
    map_val         = dbp + (pulse_pressure / 3)
    # HRV approximation: 1000/HR * variability factor (~0.04 typical)
    hrv             = round((1000 / hr) * 0.04 * hr * 0.35, 1) if hr > 0 else 35

    return {
        "Derived_BMI"            : round(bmi, 2),
        "Derived_Pulse_Pressure" : round(pulse_pressure, 1),
        "Derived_MAP"            : round(map_val, 1),
        "Derived_HRV"            : round(hrv, 1),
    }

@app.route('/health', methods=['GET'])
def health():
    return jsonify({'status': 'ok', 'model_loaded': predictor is not None})

@app.route('/predict', methods=['POST'])
def predict():
    if predictor is None:
        return jsonify({'error': 'Models not loaded.'}), 503

    data = request.get_json()
    if not data:
        return jsonify({'error': 'JSON body required'}), 400

    required = ['heart_rate', 'temperature', 'oxygen']
    for f in required:
        if f not in data:
            return jsonify({'error': f'Missing field: {f}'}), 400

    try:
        derived = derive_features(data)

        patient = {
            "Heart Rate"               : float(data.get("heart_rate", 75)),
            "Respiratory Rate"         : float(data.get("respiratory_rate", 16)),
            "Body Temperature"         : float(data.get("temperature", 37.0)),
            "Oxygen Saturation"        : float(data.get("oxygen", 98)),
            "Systolic Blood Pressure"  : float(data.get("systolic_bp", 120)),
            "Diastolic Blood Pressure" : float(data.get("diastolic_bp", 80)),
            "Age"                      : float(data.get("age", 40)),
            "Gender"                   : float(data.get("gender", 0)),
            "Weight (kg)"              : float(data.get("weight_kg", 70)),
            "Height (m)"               : float(data.get("height_m") or
                                               float(data.get("height_cm", 170)) / 100),
            "Derived_HRV"              : derived["Derived_HRV"],
            "Derived_Pulse_Pressure"   : derived["Derived_Pulse_Pressure"],
            "Derived_BMI"              : derived["Derived_BMI"],
            "Derived_MAP"              : derived["Derived_MAP"],
        }

        # Single model prediction (best model = RandomForest)
        result  = predictor.predict(patient)
        ensemble = predictor.predict_ensemble(patient)

        label = result["label"]   # "Low Risk" or "High Risk"

        # Map to 3-level risk for the frontend
        risk_level = "Normal" if label == "Low Risk" else "Emergency" \
                     if result["probability_high_risk"] >= 0.80 else "Warning"

        return jsonify({
            "prediction"           : risk_level,            # Normal/Warning/Emergency
            "raw_label"            : label,                  # Low Risk / High Risk
            "probability_high_risk": result["probability_high_risk"],
            "prob_rf"              : ensemble["prob_rf"],
            "prob_xgb"             : ensemble["prob_xgb"],
            "model_used"           : result["model_used"],
            "recommendation"       : RECOMMENDATIONS[label],
            "derived_features"     : derived,
        })

    except Exception as e:
        return jsonify({'error': f'Prediction failed: {str(e)}'}), 500

def build_patient_features(data: dict) -> dict:
    """Build the model's 14-feature dict from raw API input (shared by predict/explain)."""
    derived = derive_features(data)
    return {
        "Heart Rate"               : float(data.get("heart_rate", 75)),
        "Respiratory Rate"         : float(data.get("respiratory_rate", 16)),
        "Body Temperature"         : float(data.get("temperature", 37.0)),
        "Oxygen Saturation"        : float(data.get("oxygen", 98)),
        "Systolic Blood Pressure"  : float(data.get("systolic_bp", 120)),
        "Diastolic Blood Pressure" : float(data.get("diastolic_bp", 80)),
        "Age"                      : float(data.get("age", 40)),
        "Gender"                   : float(data.get("gender", 0)),
        "Weight (kg)"              : float(data.get("weight_kg", 70)),
        "Height (m)"               : float(data.get("height_m") or
                                           float(data.get("height_cm", 170)) / 100),
        "Derived_HRV"              : derived["Derived_HRV"],
        "Derived_Pulse_Pressure"   : derived["Derived_Pulse_Pressure"],
        "Derived_BMI"              : derived["Derived_BMI"],
        "Derived_MAP"              : derived["Derived_MAP"],
    }


@app.route('/explain', methods=['POST'])
def explain():
    """SHAP explanation for a single prediction — per-feature risk contributions."""
    if predictor is None:
        return jsonify({'error': 'Models not loaded.'}), 503

    data = request.get_json()
    if not data:
        return jsonify({'error': 'JSON body required'}), 400

    for f in ['heart_rate', 'temperature', 'oxygen']:
        if f not in data:
            return jsonify({'error': f'Missing field: {f}'}), 400

    try:
        patient = build_patient_features(data)
        explanation = predictor.explain(patient)

        label = explanation["label"]
        prob  = explanation["probability_high_risk"]
        risk_level = "Normal" if label == "Low Risk" else "Emergency" if prob >= 0.80 else "Warning"

        return jsonify({
            "prediction"           : risk_level,             # Normal/Warning/Emergency
            "raw_label"            : label,                  # Low Risk / High Risk
            "probability_high_risk": prob,
            "base_value"           : explanation["base_value"],
            "model_used"           : explanation["model_used"] + " + SHAP",
            "contributions"        : explanation["contributions"],
        })

    except Exception as e:
        return jsonify({'error': f'Explanation failed: {str(e)}'}), 500


if __name__ == '__main__':
    port = int(os.environ.get('ML_PORT', 5001))
    print(f"Starting ML service on port {port}...")
    app.run(host='0.0.0.0', port=port, debug=False)
